### PR TITLE
DIV-6627: Recording multiple HWF payments and multiple PBA and Fee account in CCD

### PR DESCRIPTION
**Pull request by madjava**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/DIV-6627

### Change description ###
Adding new `HelpWithFeesReferenceNumber` and `FeeAccountReferenceNumber` so that general referral payments and confirm service payments don't overwrite the original HWF or PBA reference numbers captured when the original digital application was made.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: definitions/divorce/json/CaseEventToFields/CaseEventToFields-deemed-and-dispensed-nonprod.json [hunk 1]
```diff
@@ -84,7 +84,7 @@
     "LiveFrom": "04/08/2020",
     "CaseTypeID": "DIVORCE",
     "CaseEventID": "ConfirmServicePayment",
-    "CaseFieldID": "SolicitorFeeAccountNumber",
+    "CaseFieldID": "FeeAccountReferenceNumber",
```
Comment (madjava) [inline definitions/divorce/json/CaseEventToFields/CaseEventToFields-deemed-and-dispensed-nonprod.json:87]: Replace so `ConfirmServicePayment` event now uses the new field

File: definitions/divorce/json/CaseEventToFields/CaseEventToFields-deemed-and-dispensed-nonprod.json [hunk 2]
```diff
@@ -110,7 +110,7 @@
     "LiveFrom": "04/08/2020",
     "CaseTypeID": "DIVORCE",
     "CaseEventID": "ConfirmServicePayment",
-    "CaseFieldID": "D8HelpWithFeesReferenceNumber",
+    "CaseFieldID": "HelpWithFeesReferenceNumber",
```
Comment (madjava) [inline definitions/divorce/json/CaseEventToFields/CaseEventToFields-deemed-and-dispensed-nonprod.json:113]: Ditto.

File: definitions/divorce/json/CaseEventToFields/CaseEventToFields-general-referral-nonprod.json [hunk 3]
```diff
@@ -133,10 +133,10 @@
     "PageDisplayOrder": 1
   },
   {
-    "LiveFrom": "20/10/2020",
+    "LiveFrom": "23/11/2020",
     "CaseTypeID": "DIVORCE",
     "CaseEventID": "generalReferralPayment",
-    "CaseFieldID": "SolicitorFeeAccountNumber",
+    "CaseFieldID": "FeeAccountReferenceNumber",
```
Comment (madjava) [inline definitions/divorce/json/CaseEventToFields/CaseEventToFields-general-referral-nonprod.json:139]: Replace so`generalReferralPayment` event now uses the new field

File: definitions/divorce/json/CaseEventToFields/CaseEventToFields-general-referral-nonprod.json [hunk 4]
```diff
@@ -157,10 +157,10 @@
     "FieldShowCondition": "GeneralReferralPaymentType=\"feeAccount\""
   },
   {
-    "LiveFrom": "20/10/2020",
+    "LiveFrom": "23/11/2020",
     "CaseTypeID": "DIVORCE",
     "CaseEventID": "generalReferralPayment",
-    "CaseFieldID": "D8HelpWithFeesReferenceNumber",
+    "CaseFieldID": "HelpWithFeesReferenceNumber",
```
Comment (madjava) [inline definitions/divorce/json/CaseEventToFields/CaseEventToFields-general-referral-nonprod.json:163]: Ditto.

File: definitions/divorce/json/CaseField/CaseField-nonprod.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  {
+    "LiveFrom": "23/11/2020",
+    "CaseTypeID": "DIVORCE",
+    "ID": "FurtherHelpWithFeesReferenceNumbers",
+    "Label": "Further HWF references",
+    "FieldType": "Collection",
+    "FieldTypeParameter": "ReferenceNumber",
+    "SecurityClassification": "Public"
+  }
+]
```
Comment (madjava) [inline definitions/divorce/json/CaseField/CaseField-nonprod.json:39]: Definitions for new reference fields and collection types for both HWF and Fee Accounts

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-nonprod.json [hunk 6]
```diff
@@ -5,5 +5,173 @@
...
+    "UserRole": "caseworker-divorce-superuser",
+    "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "23/11/2020",
+    "CaseTypeID": "DIVORCE",
+    "CaseFieldID": "FeeAccountReferenceNumber",
+    "UserRole": "caseworker-divorce-courtadmin-la",
+    "CRUD": "CRU"
+  },
```
Comment (madjava) [inline definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-nonprod.json:92]: For new HWF and FeeAccount definitions, using same RBAC as in existing D8HWF and SolFeeAccount RBAC's

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-nonprod.json [hunk 7]
```diff
@@ -5,5 +5,173 @@
...
+    "CaseFieldID": "FurtherHelpWithFeesReferenceNumbers",
+    "UserRole": "caseworker-divorce-solicitor",
+    "CRUD": "R"
+  },
+  {
+    "LiveFrom": "23/11/2020",
+    "CaseTypeID": "DIVORCE",
+    "CaseFieldID": "FurtherHelpWithFeesReferenceNumbers",
+    "UserRole": "citizen",
+    "CRUD": "R"
```
Comment (madjava) [inline definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-nonprod.json:175]: For collection types using same RBAC as in the service application collection definition

Comment (madjava) [reply definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-nonprod.json:175]: Updated to using some abbreviations as name was hitting CCD's max length

Comment (madjava) [reply definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-nonprod.json:175]: Updated to using some abbreviations as name was hitting CCD's max length

File: definitions/divorce/json/CaseField/CaseField-nonprod.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
+[
+  {
+    "LiveFrom": "23/11/2020",
+    "CaseTypeID": "DIVORCE",
+    "ID": "HelpWithFeesReferenceNumber",
+    "Label": "Help with fees reference",
+    "FieldType": "Text",
+    "RegularExpression": "([Hh][Ww][Ff]-?)?[0-9a-zA-Z]{3}-?[0-9a-zA-Z]{3}$",
```
Comment (yarpo) [inline definitions/divorce/json/CaseField/CaseField-nonprod.json:8]: why in `-nonprod` without feature flag?

Comment (yarpo) [inline definitions/divorce/json/CaseField/CaseField-nonprod.json:8]: can you please add unit tests for this regex?

File: definitions/divorce/json/CaseField/CaseField-nonprod.json [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "23/11/2020",
+    "CaseTypeID": "DIVORCE",
+    "ID": "FeeAccountReferenceNumber",
+    "Label": "Enter your account number",
+    "HintText": "Example: PBA0896366",
+    "FieldType": "Text",
+    "RegularExpression": "^([Pp][Bb][Aa])\\d{7}$",
```
Comment (yarpo) [inline definitions/divorce/json/CaseField/CaseField-nonprod.json:18]: shall we unit test those regular expressions?

File: test/unit/definitions/divorce/AuthorisationCaseField.test.js [hunk 10]
```diff
@@ -26,7 +26,8 @@ describe('AuthorisationCaseField', () => {
         'CaseField',
         'CaseField-deemed-and-dispensed-nonprod',
         'CaseField-general-email-nonprod',
-        'CaseField-general-referral-nonprod'
+        'CaseField-general-referral-nonprod',
+        'CaseField-nonprod'
```
Comment (yarpo) [inline test/unit/definitions/divorce/AuthorisationCaseField.test.js:30]: why no flag?

Comment (madjava) [reply test/unit/definitions/divorce/AuthorisationCaseField.test.js:30]: @yarpo `HelpWithFeesReferenceNumber` and  `FurtherPBAReferenceNumbers` are used in both deemed and dispensed and general referral features as they are reused. Seems the best place to me to put this.

File: definitions/divorce/json/CaseField/CaseField-nonprod.json (hunk 9 above)
Comment (madjava) [reply definitions/divorce/json/CaseField/CaseField-nonprod.json:18]: @yarpo I used the same definition that exists for `D8HelpWithFeesReferenceNumber` and `SolicitorFeeAccountNumber` as these new fields are doing same job.

File: definitions/divorce/json/CaseField/CaseField-nonprod.json (hunk 8 above)
Comment (madjava) [reply definitions/divorce/json/CaseField/CaseField-nonprod.json:8]: @yarpo I used the same definition that exists for D8HelpWithFeesReferenceNumber and SolicitorFeeAccountNumber as these new fields are doing same job.

**Review by yarpo — APPROVED** [review]

File: test/unit/definitions/divorce/AuthorisationCaseField.test.js (hunk 10 above)
Comment (madjava) [reply test/unit/definitions/divorce/AuthorisationCaseField.test.js]: Updated

File: definitions/divorce/json/CaseField/CaseField-nonprod.json (hunk 8 above)
Comment (madjava) [reply definitions/divorce/json/CaseField/CaseField-nonprod.json]: Updated

File: definitions/divorce/json/CaseField/CaseField-deemed-and-dispensed-nonprod.json [hunk 11]
```diff
@@ -136,15 +136,15 @@
     "LiveFrom": "12/08/2020",
     "CaseTypeID": "DIVORCE",
     "ID": "LabelPBANumber",
-    "Label": "Fee account number: ${SolicitorFeeAccountNumber}",
+    "Label": "Fee account number: ${FeeAccountReferenceNumber}",
```
Comment (Dizrupto) [inline definitions/divorce/json/CaseField/CaseField-deemed-and-dispensed-nonprod.json:139]: Just to confirm, General referral also will be using the new approach? :)
This field is used in both "General Referral" and "Service application" event

**Review by Dizrupto — APPROVED** [review]

File: definitions/divorce/json/CaseField/CaseField-deemed-and-dispensed-nonprod.json (hunk 11 above)
Comment (madjava) [reply definitions/divorce/json/CaseField/CaseField-deemed-and-dispensed-nonprod.json:139]: Yes, updated for both general referral payment and confirm service payment

**Pull request merged**